### DISCUSSION TITLE
New ranking rule: `words in [attribute set]`

### DISCUSSION THREAD

**Discussion by loiclec** [meilisearch/product#595] [Feedback & Feature Proposal]
We could add a new ranking rule that would sort documents by the number of words contained in a subset of the searchable attributes.
For example, let's say we have the following searchable attributes:
```json
["title", "director", "review"]
``` 
If someone has the ranking rules:
```json
["words", "typo"]
```
then the documents will be sorted first by the number of words contained in *all the searchable attributes*, without preference for any attribute in particular.

The new ranking rule `words in [attribute set]` would work similarly, except that it would be constrained to a subset of the searchable attributes. So one could have the ranking rules:
```json
[
    "words in [ title, director ]",
    "words in [ review ]",
    "typo"
]
```
and the documents will be sorted:
1. First, by the number of query words contained in `title` and `author`
2. Then, by the number of query words contained in `description`
3. Finally, by the number of typos

To illustrate how this differs from the behaviour of `words`, let's consider the following dataset:
```json
[
	{
		"title": "the lord of the rings",
		"director": "peter jackson",
		"review": "a fantastic movie"
	},
	{
		"title": "the land of hobbits",
		"director": "john smith",
		"review": "awful, this film is nothing like the lord of the rings"
	}
]
```
And the search query:
```
lord of the ring film
```

Using the first set of ranking rules, the second document would be returned first because it contains all the search query terms, while the first document would be returned later because it misses the term `film`.

Using the second set of ranking rules, the first document will be returned at the top because it contains the most query words in the attributes `[title, director]`.

Additionally, we could consider giving a weight to each attribute passed as parameter to the `words in` ranking rule, like so:
```json
[
    "words in [ title = 2.0, director = 1.5, review = 1.0 ]",
    "typo"
]
```
Such that a query word found in the attribute `title` would be worth 2 words in `review`.

Note that a weighted `words in` ranking rule would be very similar to the ranking rule `attribute` (ignoring the word positions within each attribute) and an unweighted `words in [all searchable attributes]` ranking rule would be equivalent to the current `words` ranking rule.

**Comment by gmourier** [comment]
Thanks @loiclec 👍

I'm cross-referencing a previous product discussion (https://github.com/meilisearch/product/discussions/107) where the need may be solved by this proposal if (https://github.com/meilisearch/product/discussions/593) is implemented.

**Comment by martonx** [comment]
It sounds like a must have feature!

**Comment by macraig** [comment]
Hey folks 👋 

[v1.3](https://github.com/meilisearch/meilisearch/releases/tag/v1.3.0) has been released! 🦁  You can now define fields to search on at search time ✨

**Note:** Attributes passed to `attributesToSearchOn` must be in the [searchable attributes list](https://www.meilisearch.com/docs/learn/configuration/displayed_searchable_attributes#the-searchableattributes-list).

📚 https://www.meilisearch.com/docs/reference/api/search#customize-attributes-to-search-on-at-search-time
